mongo/integration/errors_test.go

```diff
@@ -322,6 +326,126 @@ func TestErrors(t *testing.T) {
 				assert.Equal(mt, errors.Is(tc.err, matchWrapped), tc.isResult, "expected errors.Is result to be %v", tc.isResult)
 			})
 		}
+
+		mtOpts := mtest.NewOptions().MinServerVersion("4.0").Topologies(mtest.ReplicaSet)
+		mt.RunOpts("RawResponse", mtOpts, func(mt *mtest.T) {
+			mt.Run("CommandError", func(mt *mtest.T) {
+				// Mock a CommandError via failpoint with an arbitrary code.
+				mt.SetFailPoint(mtest.FailPoint{
+					ConfigureFailPoint: "failCommand",
+					Mode: mtest.FailPointMode{
+						Times: 1,
+					},
+					Data: mtest.FailPointData{
+						FailCommands: []string{"find"},
+						ErrorCode:    123,
+					},
+				})
+
+				res := mt.Coll.FindOne(context.Background(), bson.D{})
+				assert.NotNil(mt, res.Err(), "expected FindOne error, got nil")
+				se, ok := res.Err().(mongo.ServerError)
+				assert.True(mt, ok, "expected FindOne error to be ServerError")
+
+				// Assert that raw response exists and contains error code 123.
+				raw := se.RawResponse()[0]
+				assert.NotNil(mt, raw, "expected RawResponse, got nil")
+				val, err := raw.LookupErr("code")
+				assert.Nil(mt, err, "expected 'code' field in RawResponse, got %v", raw)
+				code, ok := val.AsInt64OK()
+				assert.True(mt, ok, "expected 'code' to be int64, got %v", val)
+				assert.Equal(mt, code, int64(123), "expected 'code' 123, got %d", code)
+			})
+			mt.Run("WriteError", func(mt *mtest.T) {
+				// Mock a WriteError via failpoint with an arbitrary code.
+				mt.SetFailPoint(mtest.FailPoint{
+					ConfigureFailPoint: "failCommand",
+					Mode: mtest.FailPointMode{
+						Times: 1,
+					},
+					Data: mtest.FailPointData{
+						FailCommands: []string{"delete"},
+						ErrorCode:    123,
+					},
+				})
+
+				_, err := mt.Coll.DeleteOne(context.Background(), bson.D{})
+				assert.NotNil(mt, err, "expected DeleteOne error, got nil")
```

Comment:
This suggestion assumes the above suggestion to insert duplicate `_id` is accepted.

I think the error will be a `WriteException` containing a `WriteError`.

```suggestion
				assert.NotNil(mt, err, "expected DeleteOne error, got nil")
				_, ok := err.(mongo.WriteException)
				assert.True(mt, ok, "expected type mongo.WriteException, got %v", reflect.TypeOf(err))
```

Reply:
Yep! Updated to extract the `WriteError` from the `WriteException`.